torch_xla/csrc/ops/max_pool_nd.cpp

```diff
@@ -0,0 +1,82 @@
+#include "torch_xla/csrc/ops/max_pool_nd.h"
+#include "tensorflow/compiler/xla/xla_client/debug_macros.h"
+#include "tensorflow/compiler/xla/xla_client/util.h"
+#include "torch_xla/csrc/lowering_context.h"
+#include "torch_xla/csrc/ops/infer_output_shape.h"
+#include "torch_xla/csrc/pooling.h"
+
+namespace torch_xla {
+namespace ir {
+namespace ops {
+namespace {
+
+// Infers the output shape of the max pooling operation.
+xla::Shape NodeOutputShape(
+    const Value& input, xla::int64 spatial_dim_count,
+    tensorflow::gtl::ArraySlice<const xla::int64> kernel_size,
+    tensorflow::gtl::ArraySlice<const xla::int64> stride,
+    tensorflow::gtl::ArraySlice<const xla::int64> padding) {
+  auto lower_for_shape_fn =
+      [spatial_dim_count, kernel_size, stride,
+       padding](tensorflow::gtl::ArraySlice<const xla::XlaOp> operands)
+      -> xla::XlaOp {
+    XLA_CHECK_EQ(operands.size(), 1)
+        << "Unexpected number of operands: " << operands.size();
+    return BuildMaxPoolNd(operands[0], spatial_dim_count, kernel_size, stride,
+                          padding);
+  };
+  return InferOutputShape({input.shape()}, lower_for_shape_fn);
+}
+
+c10::Symbol MaxPoolNdSymbol(xla::int64 spatial_dim_count) {
+  switch (spatial_dim_count) {
+    case 1:
+      return at::aten::max_pool1d;
+    case 2:
+      return at::aten::max_pool2d;
+    case 3:
+      return at::aten::max_pool3d;
+    default:
+      XLA_ERROR() << "Invalid number of spatial dimensions: "
+                  << spatial_dim_count;
+  }
+}
+
+}  // namespace
+
+MaxPoolNd::MaxPoolNd(const Value& input, xla::int64 spatial_dim_count,
+                     std::vector<xla::int64> kernel_size,
+                     std::vector<xla::int64> stride,
+                     std::vector<xla::int64> padding)
+    : Node(
+          ir::OpKind(MaxPoolNdSymbol(spatial_dim_count)), {input},
+          [&]() {
+            return NodeOutputShape(input, spatial_dim_count, kernel_size,
+                                   stride, padding);
+          },
+          /*num_outputs=*/1,
+          xla::util::MHash(spatial_dim_count, kernel_size, stride, padding)),
+      spatial_dim_count_(spatial_dim_count),
+      kernel_size_(kernel_size.begin(), kernel_size.end()),
```

Comment:
?
std::move to all.

Reply:
Leftover from when the parameters were slices, probably. Done.